pkg/plugins/localfiles.go

```diff
@@ -10,32 +10,49 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
-var _ fs.FS = (*LocalFS)(nil)
+var _ FS = (*LocalFS)(nil)
 
 // LocalFS is a plugins.FS that allows accessing files on the local file system.
 type LocalFS struct {
```

Comment:
I wonder if it might be better to have a separate implementation for this instead of adding this logic to the existing one 🤔 Maybe that's too weird though cc @andresmgot @marefr WDYT?

My thinking is that the existing one is quite locked down as it takes the list of files it reads on startup, but a development mode implementation should allow anything within the plugin directory - better not to mix the two concerns.

Reply:
Yeah maybe. Tricky. Wasn't really aware (forgot) of the allow list in the `LocalFS`. I guess that's security related, but we also have this https://github.com/grafana/grafana/blob/8ab7ba2b30ee793ea8d9df5a87810edf02887394/pkg/plugins/manager/signature/manifest.go#L196-L220 that wouldn't even load the plugin if it fails. For plugins in development mode I guess there's no manifest at all and therefore it just worked before?

Is the allowlist in `LocalFS` for the scenario where Grafana has loaded the plugin and verified the manifest and someone sneaks extra files into the plugin dir?

Maybe wrap the `LocalFS` in another `FS`, e.g. an `AllowFS`?

Reply:
> that wouldn't even load the plugin if it fails. For plugins in development mode I guess there's no manifest at all and therefore it just worked before?

Yes, exactly. It would cause the signature check to fail if there are extra files, but in dev mode unsigned plugins are allowed.

> Is the allowlist in LocalFS for the scenario where Grafana has loaded the plugin and verified the manifest and someone sneaks extra files into the plugin dir?

I think the allow list was just a side effect of Will's refactoring around FS abstraction, but it is good to have to protect against the case you mentioned. I like it as it provides additional security, but it slows things down in develpoment mode because it forces the developer to restart Grafana if a file is added as part of "hot-reloading" a frontend plugin while developing it (LocalFS with allow list is also used to serve plugin assets under /public/plugins/PLUGIN_ID/xyz). Esteban reported it and he was really confused by this behaviour.

> Maybe wrap the LocalFS in another FS, e.g. an AllowFS?

That's a good idea. I wanted to do it like this originally but then switched to options-based as both use a "local" filesystem, so share most of the implementation, but I like it either ways so happy to switch 😄